File: EgoCapture/Services/Upload/UploadManager.swift
import Foundation
import Combine

/// Orchestrates the full upload pipeline: chunk → upload → cleanup.
/// Singleton; manages concurrent uploads across sessions.
@MainActor
final class UploadManager: ObservableObject {

    static let shared = UploadManager()

    @Published var activeUploads: [String: UploadState] = [:]

    private var uploadTasks: [String: Task<Void, Never>] = [:]
    private var sessionReadyObserver: NSObjectProtocol?

    private init() {
        // Register the observer synchronously so new sessions are picked up immediately;
        // resuming pending uploads is deferred to keep app launch responsive (the previous
        // behaviour blocked the main actor when the user had many unfinished sessions).
        sessionReadyObserver = NotificationCenter.default.addObserver(
            forName: Notification.Name("egocaptureSessionReadyForUpload"),
            object: nil,
            queue: .main
        ) { [weak self] notification in
            guard let sessionId = notification.userInfo?["sessionId"] as? String,
                  let sessionDir = notification.userInfo?["sessionDir"] as? URL else { return }
            Task { @MainActor in
                self?.startUpload(sessionId: sessionId, sessionDir: sessionDir)
            }
        }
        Task { @MainActor [weak self] in
            // 1s breathing room — gives SwiftUI time to render first frame before we touch
            // the sessions folder + start uploads.
            try? await Task.sleep(nanoseconds: 1_000_000_000)
            self?.resumePendingUploads()
        }
    }

    // MARK: - Public API

    /// Start upload for a newly finished session.
    func startUpload(sessionId: String, sessionDir: URL) {
        guard uploadTasks[sessionId] == nil else { return }

        let config = S3Config.embedded()
        guard config.isValid else {
            print("[UploadManager] S3 credentials not configured — skipping upload for \(sessionId)")
            return
        }

        // `collectorId` here is the S3 key prefix that groups sessions together.
        // It now resolves to "<campaign>/<user-slug>" (e.g. "EgoTeste-iOS/marcos"),
        // instead of the legacy per-device UUID. The field is kept on UploadState for
        // backward compatibility with resume-in-progress data written by older builds.
        let collectorId = CampaignConfig.s3Prefix

        let task = Task {
            await runUploadPipeline(sessionId: sessionId, sessionDir: sessionDir, collectorId: collectorId, config: config)
        }
        uploadTasks[sessionId] = task
    }

    /// Retry failed uploads for a session.
    func retryUpload(sessionId: String) {
        guard uploadTasks[sessionId] == nil else { return }

        let sessionsRoot = SessionManager.shared.sessionsRoot
        let sessionDir = sessionsRoot.appendingPathComponent(sessionId)

        guard var state = UploadStateManager.load(sessionDir: sessionDir) else { return }

        let config = S3Config.embedded()
        guard config.isValid else { return }

        for i in state.files.indices where state.files[i].status == .failed {
            state.files[i].status = .pending
            state.files[i].attempts = 0
            state.files[i].lastError = nil
        }
        state.status = .uploading
        UploadStateManager.save(state, sessionDir: sessionDir)
        activeUploads[sessionId] = state

        let task = Task {
            await uploadFiles(sessionId: sessionId, sessionDir: sessionDir, config: config)
        }
        uploadTasks[sessionId] = task
    }

    /// Enqueues upload of local taxonomy JSON (`taxonomy_<session>.json` or legacy `taxonomy.json`) for sessions
    /// that already finished the pipeline before taxonomy was included in the upload manifest.
    ///
    /// Uses the session's saved `collectorId` from `upload_state.json` so keys match the original upload prefix.
    ///
    /// - Parameters:
    ///   - sessionId: If non-nil, only that session; otherwise all session folders are scanned.
    ///   - force: If true, re-uploads even when the file is already marked `.done` in state (overwrites S3).
    /// - Returns: Number of sessions for which an upload task was started.
    @discardableResult
    func backfillTaxonomyArtifacts(sessionId: String? = nil, force: Bool = false) -> Int {
        let config = S3Config.embedded()
        guard config.isValid else {
            print("[UploadManager] backfill taxonomy: S3 not configured")
            return 0
        }

        var started = 0
        for session in SessionManager.shared.listSessions() {
            if let only = sessionId, only != session.id { continue }
            guard uploadTasks[session.id] == nil else { continue }
            let sessionDir = session.directory
            guard var state = UploadStateManager.load(sessionDir: sessionDir) else { continue }
            guard let taxURL = SessionFiles.resolveExisting("taxonomy", "json", in: sessionDir) else { continue }

            let filename = taxURL.lastPathComponent
            let s3Key = "\(state.collectorId)/\(state.sessionId)/\(filename)"
            let sizeBytes = Self.localFileSize(at: taxURL)

            var shouldRun = false
            if let idx = state.files.firstIndex(where: { $0.filename == filename }) {
                switch state.files[idx].status {
                case .done:
                    if force {
                        state.files[idx].status = .pending
                        state.files[idx].attempts = 0
                        state.files[idx].lastError = nil
                        state.files[idx].completedAt = nil
                        shouldRun = true
                    }
                case .failed:
                    state.files[idx].status = .pending
                    state.files[idx].attempts = 0
                    state.files[idx].lastError = nil
                    shouldRun = true
                case .pending, .uploading:
                    break
                }
            } else {
                state.files.append(
                    UploadState.FileUploadEntry(
                        filename: filename,
                        s3Key: s3Key,
                        sizeBytes: sizeBytes,
                        status: .pending,
                        attempts: 0,
                        lastError: nil,
                        completedAt: nil
                    )
                )
                shouldRun = true
            }

            guard shouldRun else { continue }

            state.status = .uploading
            UploadStateManager.save(state, sessionDir: sessionDir)
            activeUploads[session.id] = state

            let sid = session.id
            let dir = sessionDir
            let task = Task {
                await uploadFiles(sessionId: sid, sessionDir: dir, config: config)
            }
            uploadTasks[sid] = task
            started += 1
            print("[UploadManager] backfill taxonomy: started upload for \(sid) → \(filename)")
        }
        return started
    }

    private static func localFileSize(at url: URL) -> Int64 {
        let attrs = try? FileManager.default.attributesOfItem(atPath: url.path)
        return attrs?[.size] as? Int64 ?? 0
    }

    /// Load upload state for display purposes.
    func loadState(sessionId: String) -> UploadState? {
        let sessionsRoot = SessionManager.shared.sessionsRoot
        let sessionDir = sessionsRoot.appendingPathComponent(sessionId)
        return UploadStateManager.load(sessionDir: sessionDir)
    }

    // MARK: - Pipeline

    private func runUploadPipeline(sessionId: String, sessionDir: URL, collectorId: String, config: S3Config) async {
        guard let videoURL = SessionFiles.resolveExisting("video", "mp4", in: sessionDir) else {
            print("[UploadManager] No video.mp4 found in \(sessionId)")
            uploadTasks[sessionId] = nil
            return
        }

        var state = UploadState(
            sessionId: sessionId,
            collectorId: collectorId,
            status: .chunking,
            files: [],
            createdAt: Date().timeIntervalSince1970 * 1000.0,
            lastUpdated: Date().timeIntervalSince1970 * 1000.0
        )
        UploadStateManager.save(state, sessionDir: sessionDir)
        activeUploads[sessionId] = state

        do {
            print("[UploadManager] Chunking video for session \(sessionId)...")
            let manifest = try await VideoChunkingService.chunkVideo(at: videoURL, outputDir: sessionDir)
            print("[UploadManager] Created \(manifest.totalChunks) chunks (\(String(format: "%.1f", manifest.totalDurationSec))s total)")

            state = UploadStateManager.createInitialState(
                sessionId: sessionId,
                collectorId: collectorId,
                sessionDir: sessionDir,
                chunkManifest: manifest,
                metadataFiles: SessionFiles.metadataFilenames(in: sessionDir)
            )
            UploadStateManager.save(state, sessionDir: sessionDir)
            activeUploads[sessionId] = state

        } catch {
            print("[UploadManager] Chunking failed: \(error.localizedDescription)")
            state.status = .failed
            UploadStateManager.save(state, sessionDir: sessionDir)
            activeUploads[sessionId] = state
            uploadTasks[sessionId] = nil
            return
        }

        await uploadFiles(sessionId: sessionId, sessionDir: sessionDir, config: config)
    }

    private func uploadFiles(sessionId: String, sessionDir: URL, config: S3Config) async {
        guard var state = UploadStateManager.load(sessionDir: sessionDir) else {
            uploadTasks[sessionId] = nil
            return
        }

        let uploader = S3UploadService(config: config)

        for i in state.files.indices {
            guard state.files[i].status == .pending else { continue }

            let entry = state.files[i]
            let fileURL = sessionDir.appendingPathComponent(entry.filename)

            guard FileManager.default.fileExists(atPath: fileURL.path) else {
                state.files[i].status = .failed
                state.files[i].lastError = "File not found on disk"
                UploadStateManager.save(state, sessionDir: sessionDir)
                activeUploads[sessionId] = state
                continue
            }

            state.files[i].status = .uploading
            UploadStateManager.save(state, sessionDir: sessionDir)
            activeUploads[sessionId] = state

            print("[UploadManager] Uploading \(entry.filename) → s3://\(config.bucket)/\(entry.s3Key)")

            let result = await uploader.uploadFile(localURL: fileURL, s3Key: entry.s3Key)

            switch result {
            case .success(let attempt):
                state.files[i].status = .done
                state.files[i].attempts = attempt
                state.files[i].completedAt = Date().timeIntervalSince1970 * 1000.0
                print("[UploadManager] ✓ \(entry.filename) uploaded (attempt \(attempt))")

            case .failure(let attempt, let error):
                state.files[i].status = .failed
                state.files[i].attempts = attempt
                state.files[i].lastError = error
                print("[UploadManager] ✗ \(entry.filename) failed: \(error)")
            }

            UploadStateManager.save(state, sessionDir: sessionDir)
            activeUploads[sessionId] = state
        }

        if state.isFullyUploaded {
            state.status = .completed
            print("[UploadManager] Session \(sessionId) fully uploaded — writing sentinel")
            await uploadSentinel(sessionId: sessionId, sessionDir: sessionDir, state: state, uploader: uploader)
            print("[UploadManager] Removing intermediate chunks")
            cleanupChunks(sessionDir: sessionDir, state: state)
        } else if state.hasFailures {
            state.status = .partiallyFailed
        }

        UploadStateManager.save(state, sessionDir: sessionDir)
        activeUploads[sessionId] = state
        uploadTasks[sessionId] = nil
    }

    // MARK: - Sentinel

    /// Uploads a tiny `upload_complete_{sessionId}.json` as the LAST object so the
    /// server-side MCAP builder (S3 event → Lambda) has a reliable trigger indicating
    /// that every other artifact is already in place.
    ///
    /// Failure to upload the sentinel is non-fatal for the session itself, but it
    /// does mean no MCAP will be built for this session until a manual re-run.
    private func uploadSentinel(
        sessionId: String,
        sessionDir: URL,
        state: UploadState,
        uploader: S3UploadService
    ) async {
        let filename = "upload_complete_\(sessionId).json"
        let localURL = sessionDir.appendingPathComponent(filename)
        let s3Key = "\(state.collectorId)/\(sessionId)/\(filename)"

        let payload: [String: Any] = [
            "sessionId":     sessionId,
            "collectorId":   state.collectorId,
            "completedAt":   Date().timeIntervalSince1970 * 1000.0,
            "uploadedFiles": state.completedFiles,
            "version":       1,
        ]

        guard let data = try? JSONSerialization.data(withJSONObject: payload, options: [.prettyPrinted, .sortedKeys]) else {
            print("[UploadManager] ✗ sentinel: could not encode payload")
            return
        }

        do {
            try data.write(to: localURL, options: .atomic)
        } catch {
            print("[UploadManager] ✗ sentinel: write failed — \(error.localizedDescription)")
            return
        }

        let result = await uploader.uploadFile(localURL: localURL, s3Key: s3Key)
        switch result {
        case .success(let attempt):
            print("[UploadManager] ✓ sentinel uploaded (attempt \(attempt)) → s3://\(s3Key)")
        case .failure(let attempt, let error):
            print("[UploadManager] ✗ sentinel upload failed after \(attempt) attempt(s): \(error)")
        }
    }

    // MARK: - Cleanup

    /// Remove only the intermediate `chunk_*.mp4` files after a successful upload.
    /// The original video (`video_{code}.mp4`) is always preserved locally.
    private func cleanupChunks(sessionDir: URL, state: UploadState) {
        let fm = FileManager.default
        var removed = 0
        for chunk in state.files where chunk.filename.hasPrefix("chunk_") && chunk.filename.hasSuffix(".mp4") {
            let chunkURL = sessionDir.appendingPathComponent(chunk.filename)
            if fm.fileExists(atPath: chunkURL.path) {
                try? fm.removeItem(at: chunkURL)
                removed += 1
            }
        }
        print("[UploadManager] Deleted \(removed) local chunk file(s); video preserved")
    }

    // MARK: - Resume on App Launch

    private func resumePendingUploads() {
        let config = S3Config.embedded()
        guard config.isValid else { return }

        // Disk enumeration + state parsing happens off the main actor so the UI thread
        // stays free during the scan, even if there are dozens of sessions on disk.
        Task.detached(priority: .utility) { [weak self] in
            let snapshots: [(id: String, state: UploadState, shouldRetry: Bool)] =
                SessionManager.shared.listSessions().compactMap { session in
                    guard let state = UploadStateManager.load(sessionDir: session.directory) else { return nil }
                    switch state.status {
                    case .uploading, .partiallyFailed:
                        return (session.id, state, state.pendingFiles > 0 || state.hasFailures)
                    case .completed:
                        return (session.id, state, false)
                    default:
                        return nil
                    }
                }

            guard let self else { return }
            await MainActor.run {
                for snap in snapshots {
                    self.activeUploads[snap.id] = snap.state
                }
            }

            // Serialise retries with a small gap so we do not flood the network or the
            // main actor. Each retryUpload call itself spawns a detached Task, so this
            // loop only paces the scheduling, not the actual upload work.
            for snap in snapshots where snap.shouldRetry {
                await MainActor.run { self.retryUpload(sessionId: snap.id) }
                try? await Task.sleep(nanoseconds: 200_000_000)
            }
        }
    }
}

File: android/app/src/main/java/com/kgeneye/eye/upload/UploadManager.kt
package com.kgeneye.eye.upload

import android.content.Context
import android.util.Log
import com.kgeneye.eye.session.SessionFiles
import com.kgeneye.eye.session.SessionManager
import com.kgeneye.eye.settings.AppSettings
import com.kgeneye.eye.settings.CampaignConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.io.File

/**
 * Orchestrates uploads for all locally-stored sessions. Mirrors the behaviour
 * of the iOS `UploadManager` (singleton; persists per-session state; retries
 * the queue on app launch; writes a sentinel after the last file).
 *
 * Differences from iOS:
 * - no chunking — we upload `video_<code>.mp4` as a single object (aligns with
 *   the "keep the original MP4" direction from the last product iteration).
 * - uses the AWS SDK for Kotlin, which handles SigV4 and retries at the
 *   transport layer. We keep an outer retry loop for surfacing failures and
 *   resumability across app restarts.
 */
class UploadManager private constructor(private val context: Context) {

    private val appSettings = AppSettings.get(context)
    private val sessionManager = SessionManager.get(context)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val tasks = mutableMapOf<String, Job>()
    private val tasksMutex = Mutex()

    private val _activeUploads = MutableStateFlow<Map<String, UploadState>>(emptyMap())
    val activeUploads: StateFlow<Map<String, UploadState>> = _activeUploads.asStateFlow()

    /**
     * Kick off upload for a session that just finished recording. Does nothing
     * when AWS credentials are missing (the session stays on disk for later).
     */
    fun startUpload(sessionId: String, sessionDir: File) {
        scope.launch {
            tasksMutex.withLock {
                if (tasks[sessionId] != null) return@launch
                val config = S3Config.embedded()
                if (!config.isValid) {
                    Log.w(TAG, "Embedded S3 credentials are blank — refusing to upload $sessionId")
                    return@launch
                }
                val settings = appSettings.state.value
                val collectorId = CampaignConfig.s3Prefix(settings, context)
                val job = scope.launch { runPipeline(sessionId, sessionDir, collectorId, config) }
                tasks[sessionId] = job
                job.invokeOnCompletion {
                    scope.launch { tasksMutex.withLock { tasks.remove(sessionId) } }
                }
            }
        }
    }

    /** Mark failed entries as pending and re-upload. */
    fun retryUpload(sessionId: String) {
        scope.launch {
            tasksMutex.withLock {
                if (tasks[sessionId] != null) return@launch
                val sessionDir = File(sessionManager.sessionsRoot, sessionId)
                val state = UploadStateStore.load(sessionDir) ?: return@launch
                val config = S3Config.embedded()
                if (!config.isValid) return@launch

                state.files.forEach { entry ->
                    if (entry.status == UploadState.FileEntry.FileStatus.failed ||
                        entry.status == UploadState.FileEntry.FileStatus.uploading
                    ) {
                        entry.status = UploadState.FileEntry.FileStatus.pending
                        entry.attempts = 0
                        entry.lastError = null
                    }
                }
                state.status = UploadState.SessionStatus.uploading
                UploadStateStore.save(state, sessionDir)
                publish(state)

                val job = scope.launch { uploadFiles(sessionId, sessionDir, config) }
                tasks[sessionId] = job
                job.invokeOnCompletion {
                    scope.launch { tasksMutex.withLock { tasks.remove(sessionId) } }
                }
            }
        }
    }

    /** Scan disk on launch and resume any session that was mid-upload. */
    fun resumePendingUploads() {
        scope.launch {
            val config = S3Config.embedded()
            if (!config.isValid) return@launch
            delay(500)
            for (session in sessionManager.listSessions()) {
                val state = UploadStateStore.load(session.directory) ?: continue
                publish(state)
                val needsWork = state.status == UploadState.SessionStatus.uploading ||
                        state.status == UploadState.SessionStatus.partiallyFailed
                if (needsWork) retryUpload(session.id)
                delay(200)
            }
        }
    }

    fun loadState(sessionId: String): UploadState? {
        val dir = File(sessionManager.sessionsRoot, sessionId)
        return UploadStateStore.load(dir)
    }

    // region pipeline

    private suspend fun runPipeline(
        sessionId: String,
        sessionDir: File,
        collectorId: String,
        config: S3Config,
    ) {
        val state = UploadStateStore.load(sessionDir)
            ?: UploadStateStore.initialState(sessionId, collectorId, sessionDir).also {
                UploadStateStore.save(it, sessionDir)
            }
        publish(state)
        uploadFiles(sessionId, sessionDir, config)
    }

    private suspend fun uploadFiles(sessionId: String, sessionDir: File, config: S3Config) {
        val state = UploadStateStore.load(sessionDir) ?: return
        val uploader = S3UploadService(config)

        // A process can die while a large file (usually video) is marked as
        // `uploading`. On the next run there is no active network request, so
        // that persisted state must be treated as resumable work.
        var resetStaleUploading = false
        state.files.forEach { entry ->
            if (entry.status == UploadState.FileEntry.FileStatus.uploading) {
                entry.status = UploadState.FileEntry.FileStatus.pending
                resetStaleUploading = true
            }
        }
        if (resetStaleUploading) {
            UploadStateStore.save(state, sessionDir)
            publish(state)
        }

        for (entry in state.files) {
            if (entry.status != UploadState.FileEntry.FileStatus.pending) continue
            val local = File(sessionDir, entry.filename)
            if (!local.exists()) {
                entry.status = UploadState.FileEntry.FileStatus.failed
                entry.lastError = "File not found on disk"
                UploadStateStore.save(state, sessionDir); publish(state)
                continue
            }

            entry.status = UploadState.FileEntry.FileStatus.uploading
            UploadStateStore.save(state, sessionDir); publish(state)

            Log.i(TAG, "Uploading ${entry.filename} → s3://${config.bucket}/${entry.s3Key}")
            when (val res = uploader.uploadFile(local, entry.s3Key)) {
                is S3UploadService.Result.Success -> {
                    entry.status = UploadState.FileEntry.FileStatus.done
                    entry.attempts = res.attempt
                    entry.completedAt = System.currentTimeMillis().toDouble()
                }
                is S3UploadService.Result.Failure -> {
                    entry.status = UploadState.FileEntry.FileStatus.failed
                    entry.attempts = res.attempt
                    entry.lastError = res.error
                }
            }
            UploadStateStore.save(state, sessionDir); publish(state)
        }

        when {
            state.isFullyUploaded -> {
                state.status = UploadState.SessionStatus.completed
                writeSentinel(sessionId, sessionDir, state, uploader)
            }
            state.hasFailures -> state.status = UploadState.SessionStatus.partiallyFailed
        }
        UploadStateStore.save(state, sessionDir); publish(state)
    }

    @OptIn(ExperimentalSerializationApi::class)
    private suspend fun writeSentinel(
        sessionId: String,
        sessionDir: File,
        state: UploadState,
        uploader: S3UploadService,
    ) {
        val filename = "upload_complete_$sessionId.json"
        val localFile = File(sessionDir, filename)
        val s3Key = "${state.collectorId}/$sessionId/$filename"

        val payload = buildJsonObject {
            put("sessionId", sessionId)
            put("collectorId", state.collectorId)
            put("completedAt", System.currentTimeMillis().toDouble())
            put("uploadedFiles", state.completedFiles)
            put("version", 1)
        }
        val json = Json { prettyPrint = true; prettyPrintIndent = "  " }
        localFile.writeText(json.encodeToString(kotlinx.serialization.json.JsonObject.serializer(), payload))

        when (val res = uploader.uploadFile(localFile, s3Key)) {
            is S3UploadService.Result.Success -> Log.i(TAG, "Sentinel uploaded for $sessionId (attempt ${res.attempt})")
            is S3UploadService.Result.Failure -> Log.w(TAG, "Sentinel upload failed for $sessionId: ${res.error}")
        }
    }

    private fun publish(state: UploadState) {
        _activeUploads.value = _activeUploads.value.toMutableMap().apply {
            this[state.sessionId] = state
        }
    }

    // endregion

    companion object {
        private const val TAG = "UploadManager"

        @Volatile private var instance: UploadManager? = null
        fun get(context: Context): UploadManager = instance ?: synchronized(this) {
            instance ?: UploadManager(context.applicationContext).also { instance = it }
        }
    }
}
